### ISSUE TITLE
parameter: define task with unused parameter

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-flow#3447]
#### What

If you define a task in `[runtime]` using a parameter from outside of the permitted range Cylc will fail validation (and run).

This is inconsistent with the behaviour of un-parameterised tasks and families where you can define what you like in `[runtime]` whether of not these tasks or families are actually used in the graph.

#### Justification

A good use for manually specifying parameter values is to handle edge cases e.g:

```ini
[cylc]
    [[parameters]]
         foo= a, b, c

[runtime]
    [[task<foo>]]
        # common settings

    [[task<foo=a>]]
        # edge case
        [[[environment]]]
            FOO = 1
```

The edge case of `foo=a` is now formally handled in the suite. The issue comes when `a` isn't in `foo`:

```ini
[cylc]
    [[parameters]]
         foo= b, c, d
```

```console
$ cylc validate .
ParamExpandError: parameter foo out of range: foo=a
```

The solution is ugly:

```ini
#!Jinja2

{% set FOO = ["b", "c", "d"] %}

[cylc]
    [[parameters]]
        foo = {{ FOO | join(", ") }}

[runtime]
    [[task<foo>]]
        # common settings

{% if "a" in FOO %}
    [[task<foo=a>]]
        # edge case
        [[[environment]]]
            FOO = 1
{% endif %}
```

Users should have to programaticaly disable parts of their runtime just because the current suite configuration doesn't use it. This can get quite difficult to handle when inheritance is involved.

#### Proposal

```ini
[cylc]
     [[parameters]]
        foo = b, c, d

[scheduling]
    [[graph]]
        R1 = task<foo=a>  # ParamExpandError: parameter foo out of range: foo=a

[runtime]
    [[task<foo=a>]]  # No error
```

**Comment by hjoliver** [comment]
@oliver-sanders - I can see your point, but I think it is arguable.

If an illegal parameter value is selected, it seems more natural to me to assume an error (i.e. that the user intended to select a legal value, but made a typo). 

For unparameterized tasks, if you make a typo in the task name under runtime it creates an "unused" task definition that is ignored (as you point out) but it also creates a naked dummy task in the graph which is not ignored (with `--strict` at least). 

With parameters you won't get naked dummy task detection as insurance against a genuine error (if only the general case `task<foo>` is used in the graph).

So it's not entirely obvious to me which way we should go here.

**Comment by kinow** [comment]
Maybe we could have a strict mode, enabled by default. Though I agree for the risk of having typos in parameters I wouldn't use this feature myself.

**Comment by oliver-sanders** [comment]
The typo argument is fair, however, it's the opposite way around to the "naked dummy task" scenario.

Perhaps there's a syntax solution?

```ini
[cylc]
     [[parameters]]
        foo = b, c, d

[scheduling]
    [[graph]]
        R1 = task<foo=a>  # ParamExpandError: parameter foo out of range: foo=a

[runtime]
    [[task<foo=a>]]  # ParamExpandError: parameter foo out of range: foo=a
    [[task<foo~a>]]  # No error
```

**Comment by hjoliver** [comment]
Yeah I like the syntax suggestion.  With that, users can mark certain cases as deliberate, much easier than the Jinja2 cut-out hack.

**Comment by kinow** [comment]
+1 for a different syntax

**Comment by hjoliver** [comment]
Would it be mad to extend your syntax idea to all tasks?  i.e. `~fish` under runtime defines task `fish` but don't worry if there is no such task in the workflow graph.  The parameter case could then be `~task<foo=a>`, I guess.

**Comment by oliver-sanders** [comment]
Cylc doesn't complain if there is no such task in the graph though?

**Comment by hjoliver** [comment]
> Cylc doesn't complain if there is no such task in the graph though?

It doesn't currently, but if we adopted this syntax (I suggest) it would. i.e. complain (fail validation) if an "unused" task definition is found, unless it uses the tilde syntax.

**Comment by hjoliver** [comment]
(Put to milestone *someday* ... we can bump forward if time allows)
